### DISCUSSION TITLE
Error Implementing Scripts

### DISCUSSION THREAD

**Discussion by jacen2110** [netbox-community/netbox#11843] [Help Wanted!] [answered]
I recently started using Netbox and am trying to load scripts, but when I did this is the error I get. I thought I restarted the WSGI, but I am not totally sure. How do you restart it exactly?:

WSGI service not restarted after upgrade - If this installation has recently been upgraded, check that the WSGI service (e.g. gunicorn or uWSGI) has been restarted. This ensures that the new code is running.

The complete exception is provided below:

<class 'ModuleNotFoundError'>

No module named 'pynetbox'

Python version: 3.8.10
NetBox version: 3.4.5

**Comment by candlerb** [comment, marked as answer]
When you say "trying to load scripts", what exactly do you mean?  Did you put some code in the /opt/netbox/netbox/scripts/ directory?

If so, then it seems your custom script is trying to use a module "pynetbox" but you don't have this available in your netbox virtualenv.  Try adding pynetbox to `local_requirements.txt` then re-running `upgrade.sh` to rebuild the virtualenv.  As an aside though: it would be very unusual for a custom script to use pynetbox, when it has direct access to the Django models and can access the database safely using transactions (this allows the custom script to have a dry run which is rolled back)

If that's not the case, then please can you explain more clearly what you're doing, and show the code.

**Reply by jacen2110** [reply]
Yes, that is what I meant. I was just trying to see if I could put some of them into my local netbox. I apologize for the lack of info, I am very new at this. 

This is what I was following as a first trial run at scripts in Netbox:
https://github.com/nwiecz/Netbox-ipscanner

**Reply by candlerb** [reply]
Ah right. It does say [there](https://github.com/nwiecz/Netbox-ipscanner#requirement) to add the requirements to `local_requirements.txt` - but it doesn't tell you to rebuild your virtualenv.

If you don't want to rebuild your virtualenv entirely, you can just add the new packages into it:

```
cd /opt/netbox
source venv/bin/activate
pip install -r local_requirements.txt
deactivate
```

You may need `sudo -s` first to get a root shell.

**Reply by jacen2110** [reply]
Thank you. But now there seems to be an error in the script code itself haha... Not sure if I should just delete the script and try a different on or not. The error I get now is:
<class 'NameError'>

name 'BooleanVar' is not defined

**Reply by candlerb** [reply]
Maybe script is broken.  Try adding

```
from extras.scripts import BooleanVar
```

**Reply by jacen2110** [reply]
Yeah, I tried that. Asked ChatGPT for some help and it suggested two things and neither of them fixed it unfortunately. I am not sure why.

**Replaced this**: from extras.scripts import Script
**with this**: from extras.scripts.forms import BooleanVar, StringVar, Script

And that didn't solve it so I

**Replaced this**: import pynetbox, urllib3, networkscan, socket, ipaddress
**with this**: import pynetbox, urllib3, network_scan, socket, ipaddress

**Reply by candlerb** [reply]
> Asked ChatGPT for some help and it suggested two things and neither of them fixed it unfortunately

Sigh. ChatGPT is not intelligent: it spews out stuff which might have been said by someone in the past (or is a combination of things that other people said in the past).  The stuff it generates looks plausible, sometimes even authoritative, but is often wrong.

**Reply by jacen2110** [reply]
Yeah, I have found that to be true but it is better than I at scripting (lol)

**Reply by lasag-2511** [reply]
Hi i have the same issue ModuleNotFoundError: No module named 'extras.scripts'
did you try to find a way to solve it

**Reply by candlerb** [reply]
It's very likely not the same issue.

Start a new thread. In that thread, show your Netbox version, the actual code you're trying to run, how you're running it (e.g. is it a custom script being run from the web interface, a report being run from the web interface, a standalone script etc), and the full backtrace.

**Reply by lasag-2511** [reply]
Thanks for the reply. https://github.com/bbird81/Netbox-ipscanner/issues/9 - issue already opened
